src/transformStateWithClones.js

```diff
@@ -6,8 +6,46 @@
  *
  * @return {Object[]}
  */
+
 function transformStateWithClones(state, actions) {
-  // write code here
+  const allStates = [];
+  const realState = Object.assign({}, state);
+
+  for (const action of actions) {
+    switch (action.type) {
+      case 'addProperties':
+        addProperties(realState, action.extraData);
+        break;
+
+      case 'removeProperties':
+        removeProperties(realState, action.keysToRemove);
+        break;
+
+      case 'clear':
+        clearProperties(realState);
+        break;
+    }
+
+    allStates.push(Object.assign({}, realState));
+  }
+
+  return allStates;
+}
+
+function addProperties(state, extraData) {
+  Object.assign(state, extraData);
+}
+
+function removeProperties(state, keysToRemove) {
+  for (const key of keysToRemove) {
+    delete state[key];
+  }
+}
+
+function clearProperties(state) {
+  for (const key in state) {
+    delete state[key];
+  }
 }
```

Comment:
These helper functions `addProperties`, `removeProperties`, and `clearProperties` are directly modifying the `state` parameter, which goes against the checklist item [CODE STYLE] that advises not to mutate objects. Instead, these functions should create and return a new copy of the state with the applied changes.